### ISSUE TITLE
Kernel/Kernel Matrix Computation

### ISSUE THREAD

**Issue by trthatcher** [trthatcher/MLKernels.jl#36]
@st-- 

So I was thinking a bit about the approach to the kernel/kernel matrix computation.

If you have a function f:RxR -> R, then you can apply it element-wise to two vectors and sum the results. So for x = [x1,x2] and y = [y1,y2] the dot product the function would be f(x,y) = xy. Then k(x,y) = f(x1,y1) + f(x2,y2) = x1y1 + x2y2. The squared distance is just f(x,y) = (x-y)^2

If f is a valid positive or negative definite kernel in RxR, then we have a way to extend the kernel to R^n x R^2 by summing the element-wise results.

I was thinking we could take a more modular approach. Technically, the kernels we have now are a composition of a positive definite kernel (the polynomial kernel takes the dot product) or a negative definite positive-valued kernel (the euclidean distance in what we have implemented) and we could define these 'input' kernels in terms of `f`.

I was thinking we could abstract `kernelmatrix` into two functions: generic `kernelize` (the way it currently operates) and a generic `pairwise` which operates on two vectors/matrices. We would create a new class of kernels and `pairwise` would dispatch on those types. Examples of those new base kernels would be `SquaredDistance` and `ScalarProduct`. However, we can easily extend it:
- `f(x,y) = sin(x-y)^2` (sine squared kernel - also negative definite)
- `f(x,y) = (x-y)^2/(x+y)` (chi squared kernel for R+ x R+)

These can all be extended naturally to cover the periodic kernel weights... I was thinking two weight vectors and `f` could be defined like this:

k(x,y) = u1_sin(w1_x1 - w1_y1)^2 + u2_sin(w2_x2 - w2_y2)^2

Where u and w are your weight vectors. (I know this could add redundancy - just an idea)

Long story short, three levels:
- 'Base' simple kernels that will be defined for `pairwise` - ARD would be defined at this level
- Derived kernels that are a function of the matrix that pairwise returns
- Composite kernels as we have them now.

Anyway, I think an approach along these lines would also give us a nice modular approach to the derivatives. We can always ensure there's generic fall-back methods basically exactly how we have them now. Let me know your thoughts
